core/kernel/notif.c

```diff
@@ -106,6 +123,273 @@ void notif_send_async(uint32_t value)
+
+	if (notif)
+		set_it_mask(id, do_mask);
+}
+
+/*
+ * Activate (enable) or deactivate (disable) an interrupt notification
+ *
+ * @id Interrupt identifier
+ * @do_enable True to enable the event detection, false disable it
+ */
+TEE_Result notif_it_set_state(unsigned int id, bool do_enable)
+{
+	struct notif_it *notif = find_notif_it(id);
+
+	if (!notif)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	set_it_mask(id, !do_enable);
+
+	if (notif->ops && notif->ops->set_state)
+		return notif->ops->set_state(notif, do_enable);
+
+	/* Notifier may not implement set_state */
+	return TEE_SUCCESS;
+}
+
+/*
+ * Enable or disable the low power wakeup capability of the interrupt
+ *
+ * @id Interrupt identifier
+ * @do_enable True to enable wakeup support, false to disable it
+ */
+TEE_Result notif_it_set_wakeup(unsigned int id, bool do_enable)
+{
+	struct notif_it *notif = find_notif_it(id);
+
+	if (!notif)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (notif->ops && notif->ops->set_wakeup)
+		return notif->ops->set_wakeup(notif, do_enable);
+
+	/* Notifier must implement set_wakeup for a wakeup source interrupt */
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
+/*
+ * Notify an interrupt event to normal world
+ *
+ * @notif_it Reference to registered notifier
+ */
+void notif_it_raise_event(struct notif_it *notif_it)
+{
+	uint32_t old_itr_status = 0;
+
+	if (find_notif_it(notif_it->id) != notif_it) {
+		DMSG("Invalid notif handle");
+		return;
+	}
+
+	old_itr_status = cpu_spin_lock_xsave(&it_lock);
+
+	bit_set(notif_it_pending, notif_it->id);
+
+	if (!bit_test(notif_it_masked, notif_it->id))
+		itr_raise_pi(CFG_CORE_ASYNC_NOTIF_GIC_INTID);
+
+	cpu_spin_unlock_xrestore(&it_lock, old_itr_status);
+}
+
+/*
+ * Register an interrupt notifier
+ *
+ * @id The interrupt number/ID
+ * @ops Operations for that interrupt notifier
+ * @notif_it Preallocated instance address or output reference
+ *   Input: if not NULL and pointing to a non NULL cell, this is
+ *   the preallocated struct notif_it to be used.
+ *   Input: if not NULL and pointing to NULL, struct notif_it
+ *   instance must be allocated.
+ *   Output: if not NULL, points to registered struct notif_it
+ *
+ * Return a TEE_Result compliant code
+ */
+TEE_Result notif_it_register(unsigned int id, const struct notif_it_ops *ops,
+			     void *priv, struct notif_it **notif_it_ref)
+{
+	struct notif_it *notif = NULL;
+	uint32_t old_itr_status = 0;
+
+	if (id > NOTIF_IT_VALUE_MAX)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (find_notif_it(id))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	assert(!bit_test(notif_it_pending, id));
+	old_itr_status = cpu_spin_lock_xsave(&it_lock);
+	bit_clear(notif_it_pending, id);
+	cpu_spin_unlock_xrestore(&it_lock, old_itr_status);
+
+	if (notif_it_ref && *notif_it_ref) {
+		notif = *notif_it_ref;
+		memset(notif, 0, sizeof(*notif));
+		if (!is_unpaged(notif)) {
+			DMSG("notif instance must be unpaged");
+			return TEE_ERROR_GENERIC;
+		}
+	} else {
+		notif = calloc(1, sizeof(*notif));
```

Comment:
I don't think we need two ways of registering an interrupt notifier.

Reply:
+1